- msbuild
    -  Microsoft Build Engine是一個自由且開源的構建工具集，適用於受控代碼以及原生C++代碼，是.NET框架的一部分

- SCons
    - SCons，一種軟體開發工具程式，功能類似於UNIX上的make、autoconf與automake工具


Team calendar

- 在 macOS 上卸载 Docker
    - https://www.jiyik.com/tm/xwzj/opersys_4412.html


https://github.com/jdittrich/quickMockup




>>> r = requests.post(url, headers=headers, json={'farmID': '10128048062464'})
Traceback (most recent call last):
  File "D:\python3\Lib\site-packages\urllib3\connectionpool.py", line 467, in _make_request
    self._validate_conn(conn)
  File "D:\python3\Lib\site-packages\urllib3\connectionpool.py", line 1099, in _validate_conn
    conn.connect()
  File "D:\python3\Lib\site-packages\urllib3\connection.py", line 653, in connect
    sock_and_verified = _ssl_wrap_socket_and_match_hostname(
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\python3\Lib\site-packages\urllib3\connection.py", line 806, in _ssl_wrap_socket_and_match_hostname
    ssl_sock = ssl_wrap_socket(
               ^^^^^^^^^^^^^^^^
  File "D:\python3\Lib\site-packages\urllib3\util\ssl_.py", line 465, in ssl_wrap_socket
    ssl_sock = _ssl_wrap_socket_impl(sock, context, tls_in_tls, server_hostname)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\python3\Lib\site-packages\urllib3\util\ssl_.py", line 509, in _ssl_wrap_socket_impl
    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\python3\Lib\ssl.py", line 455, in wrap_socket
    return self.sslsocket_class._create(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\python3\Lib\ssl.py", line 1046, in _create
    self.do_handshake()
  File "D:\python3\Lib\ssl.py", line 1321, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: DH_KEY_TOO_SMALL] dh key too small (_ssl.c:1000)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "D:\python3\Lib\site-packages\urllib3\connectionpool.py", line 793, in urlopen
    response = self._make_request(
               ^^^^^^^^^^^^^^^^^^^
  File "D:\python3\Lib\site-packages\urllib3\connectionpool.py", line 491, in _make_request
    raise new_e
urllib3.exceptions.SSLError: [SSL: DH_KEY_TOO_SMALL] dh key too small (_ssl.c:1000)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "D:\python3\Lib\site-packages\requests\adapters.py", line 486, in send
    resp = conn.urlopen(
           ^^^^^^^^^^^^^
  File "D:\python3\Lib\site-packages\urllib3\connectionpool.py", line 847, in urlopen
    retries = retries.increment(
              ^^^^^^^^^^^^^^^^^^
  File "D:\python3\Lib\site-packages\urllib3\util\retry.py", line 515, in increment
    raise MaxRetryError(_pool, url, reason) from reason  # type: ignore[arg-type]
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='emng-test.zoomlion.com', port=443): Max retries exceeded with url: /open-sharing-platform/zlapi/irri/dfzg/LiquidLevel (Caused by SSLError(SSLError(1, '[SSL: DH_KEY_TOO_SMALL] dh key too small (_ssl.c:1000)')))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "D:\python3\Lib\site-packages\requests\api.py", line 115, in post
    return request("post", url, data=data, json=json, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\python3\Lib\site-packages\requests\api.py", line 59, in request
    return session.request(method=method, url=url, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\python3\Lib\site-packages\requests\sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\python3\Lib\site-packages\requests\sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\python3\Lib\site-packages\requests\adapters.py", line 517, in send
    raise SSLError(e, request=request)
requests.exceptions.SSLError: HTTPSConnectionPool(host='emng-test.zoomlion.com', port=443): Max retries exceeded with url: /open-sharing-platform/zlapi/irri/dfzg/LiquidLevel (Caused by SSLError(SSLError(1, '[SSL: DH_KEY_TOO_SMALL] dh key too small (_ssl.c:1000)')))



pip install fcntl
ERROR: Could not find a version that satisfies the requirement fcntl (from versions: none)
ERROR: No matching distribution found for fcntl
  - Unfortunately Gunicorn is for a UNIX environment and is incompatible with Windows. stackoverflow.com/a/11087777/12362709 – 


```
$ pip3 install spdx-tools==0.8.0
ERROR: Could not find a version that satisfies the requirement spdx-tools==0.8.0 (from versions: 0.3, 0.4.1, 0.5.1, 0.5.2, 0.5.3, 0.5.4, 0.6.0, 0.6.1, 0.7.0a2, 0.7.0a3, 0.7.0rc0, 0.7.0)
ERROR: No matching distribution found for spdx-tools==0.8.0

----------
Root cause:
Current Python version is 3.6, should upgrade to 3.8

How to create virtualenv with py3.8 ?

```

### Vmware
- vSphere
- vSan
- esx, esxi, NFS, ISCSI 
    - esxi
    ```
    VMware ESXi是VMware开发的企业级类型 I（type-1）Hypervisor，用于硬件虚拟化。作为类型 I（type-1）Hypervisor，ESXi不是安装在操作系统上的软件应用程序，而是直接安装在硬件上并且集成了重要的操作系统组件，如内核[4]。VMware ESXi原名ESX，2010年ESX4.1版本发布后，VMware将ESX更名为ESXi[5]。
    ```
- vmware-mount

有时候我们可能需要使用保存在虚拟机中的某个文件，但是如果要启动Vmware Workstation，然后启动虚拟机，然后再复制文件，这样做未免太麻烦了些。这里要介绍前面提供的第一个小工具了，Vmware mount，该工具可以把虚拟机的硬盘文件映射为宿主机上的硬盘分区。

- vmware-vdiskmanager
- VMware Skyline
什么是 VMware Skyline？
答：VMware Skyline 是一种主动的自助支持技术，持有有效的 Production 支持或 Premier 服务合同的客户可享用。Skyline 可自动并安全地收集、汇总并分析客户特定产品使用情况的数据，以主动发现潜在问题并更快地解决问题。

- OVF
    - `Deploy VM using OVF template`


### photon OS


- bitnami
    - 

- Tanzu


### news
- VMware全面拥抱K8s，三大策略打造全新Tanzu产品线 https://www.10lun.com/article/41190.html


- What's stream ?


- nginx $request_filename
    - 

- cgi & fastcgi


- merge_hooks

- stop_after_attempt


- Markdown anchor
  - https://stackoverflow.com/questions/5319754/cross-reference-named-anchor-in-markdown


https://github.com/joeltennant/Jekyll-and-Docker-Compose
```shell
# docker-compose up
[+] Running 2/2
 ⠿ Network jekyll-and-docker-compose_default     Created                                                                                                                                               0.0s
 ⠿ Container jekyll-and-docker-compose-jekyll-1  Created                                                                                                                                               0.1s
Attaching to jekyll-and-docker-compose-jekyll-1
jekyll-and-docker-compose-jekyll-1  | ruby 3.1.1p18 (2022-02-18 revision 53f5fc4236) [x86_64-linux-musl]
jekyll-and-docker-compose-jekyll-1  |   Logging at level: debug
jekyll-and-docker-compose-jekyll-1  |     Jekyll Version: 4.2.2
jekyll-and-docker-compose-jekyll-1  | Configuration file: /srv/jekyll/_config.yml
jekyll-and-docker-compose-jekyll-1  |   Logging at level: debug
jekyll-and-docker-compose-jekyll-1  |     Jekyll Version: 4.2.2
jekyll-and-docker-compose-jekyll-1  |                     ------------------------------------------------
jekyll-and-docker-compose-jekyll-1  |       Jekyll 4.2.2   Please append `--trace` to the `serve` command 
jekyll-and-docker-compose-jekyll-1  |                      for any additional information or backtrace. 
jekyll-and-docker-compose-jekyll-1  |                     ------------------------------------------------
jekyll-and-docker-compose-jekyll-1  | /usr/local/lib/ruby/3.1.0/fileutils.rb:243:in `mkdir': Permission denied @ dir_s_mkdir - /srv/jekyll/.jekyll-cache (Errno::EACCES)
jekyll-and-docker-compose-jekyll-1  |   from /usr/local/lib/ruby/3.1.0/fileutils.rb:243:in `fu_mkdir'
jekyll-and-docker-compose-jekyll-1  |   from /usr/local/lib/ruby/3.1.0/fileutils.rb:221:in `block (2 levels) in mkdir_p'
jekyll-and-docker-compose-jekyll-1  |   from /usr/local/lib/ruby/3.1.0/fileutils.rb:219:in `reverse_each'
jekyll-and-docker-compose-jekyll-1  |   from /usr/local/lib/ruby/3.1.0/fileutils.rb:219:in `block in mkdir_p'
jekyll-and-docker-compose-jekyll-1  |   from /usr/local/lib/ruby/3.1.0/fileutils.rb:211:in `each'
jekyll-and-docker-compose-jekyll-1  |   from /usr/local/lib/ruby/3.1.0/fileutils.rb:211:in `mkdir_p'
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/cache.rb:184:in `dump'
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/cache.rb:101:in `[]='
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/cache.rb:45:in `clear_if_config_changed'
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:118:in `reset'
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/site.rb:35:in `initialize'
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/commands/build.rb:30:in `new'
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/commands/build.rb:30:in `process'
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/command.rb:91:in `block in process_with_graceful_fail'
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/command.rb:91:in `each'
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/command.rb:91:in `process_with_graceful_fail'
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/gems/jekyll-4.2.2/lib/jekyll/commands/serve.rb:86:in `block (2 levels) in init_with_program'
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `block in execute'
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `each'
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/gems/mercenary-0.4.0/lib/mercenary/command.rb:221:in `execute'
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/gems/mercenary-0.4.0/lib/mercenary/program.rb:44:in `go'
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/gems/mercenary-0.4.0/lib/mercenary.rb:21:in `program'
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/gems/jekyll-4.2.2/exe/jekyll:15:in `<top (required)>'
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/bin/jekyll:25:in `load'
jekyll-and-docker-compose-jekyll-1  |   from /usr/gem/bin/jekyll:25:in `<main>'
jekyll-and-docker-compose-jekyll-1 exited with code 1
```


- Splunk
    - https://www.youtube.com/watch?v=m95GiTF0zd0
    - https://www.youtube.com/watch?v=bO_-fv6e7u4
    - https://medium.com/airwalk/log-aggregation-in-kubernetes-and-transporting-logs-to-splunk-for-analysis-ad8599607372
    - https://cloud.google.com/architecture/logging-anthos-with-splunk-connect?hl=zh-cn
    - k8s logs
        - objects
        - metrics


setup a splunk server:
```shell
docker run --name splunk -p 8000:8000 -p 8088:8088 -d outcoldman/splunk:6.3.3


docker run --name hello --log-driver=splunk --log-opt splunk-token=022F1FCE-B904-4E0E-B0A1-EB4492F61B9D --log-opt splunk-url=http://139.196.39.92:8088 --log-opt splunk-sourcetype=Docker rickfast/hello-oreilly

docker run --name hello --log-driver=splunk --log-opt splunk-token=1620a639-5064-43dc-8d81-72ae38ec639b --log-opt splunk-url=http://10.79.128.59:8088 --log-opt splunk-sourcetype=Docker rickfast/hello-oreilly
```






```shell
curl -k https://139.196.39.92:8088/services/collector/event -H "Authorization: Splunk C8D04ABF-AE0A-4FEB-935C-9F0751FEA816" -d '{"event": "hello world"}'
```

splunk log success: https://www.youtube.com/watch?v=qROXrFGqWAU&t=11s

curl https://10.79.128.59:8088/services/collector/event -H "Authorization: Splunk 1620a639-5064-43dc-8d81-72ae38ec639b" -d '{"event": "hello world"}'

docker built-in send log to splunk ?
https://www.w3cschool.cn/doc_docker_1_13/docker_1_13-engine-admin-logging-splunk-index.html

install splunk by cmd
https://www.inmotionhosting.com/support/security/install-splunk/

9ED0A79E-F7B8-43DC-B7A0-7B49AE7450B9

```shell
root@iZuf6bpc1lt9qlf2ma9p2lZ:~# helm install anthos-splunk -f values.yaml --namespace splunk-connect https://github.com/splunk/splunk-connect-for-kubernetes/releases/download/1.4.1/splunk-connect-for-kubernetes-1.4.1.tgz

Error: INSTALLATION FAILED: Get "https://github.com/splunk/splunk-connect-for-kubernetes/releases/download/1.4.1/splunk-connect-for-kubernetes-1.4.1.tgz": unexpected EOF
```

```shell
$ kubectl create namespace splunk-connect
$ kubectl config set-context --current --namespace=splunk-connect
$ 
helm install anthos-splunk -f values.yaml --namespace splunk-connect https://github.com/splunk/splunk-connect-for-kubernetes/releases/download/1.4.1/splunk-connect-for-kubernetes-1.4.1.tgz
```


```shell
   58  sudo apt-get install libmysqlclient-dev
   59  sudo apt-get install mysql-server
   60  pip3 install  mysqlclient==2.0.3
```

- https://superuser.com/questions/497449/saving-large-webpage-as-image
https://erd.dbdesigner.net/designer/schema/guest_template
